staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -674,7 +675,221 @@ type crdConversionRESTOptionsGetter struct {
+	u.UnstructuredContent()["metadata"] = map[string]interface{}{}
+	u.SetKind(kind)
+	u.SetAPIVersion(apiVersion)
+	setObjectMeta(u, objectMeta)
+
+	return nil
+}
+
+// dropInvalidTypedFields removes fields with wrong types, avoiding them creating more harm.
+// Only list those fields where it is nore important to keep the objects available to the user
+// than preserving all (invalid) fields.
+func dropInvalidTypedFields(u *unstructured.Unstructured) {
+	metadataRaw := u.UnstructuredContent()["metadata"]
+	if metadataRaw == nil {
+		return
+	}
+
+	metadata, ok := metadataRaw.(map[string]interface{})
+	if !ok {
+		unstructured.RemoveNestedField(u.UnstructuredContent(), "metadata")
+		return
+	}
+
+	if _, _, err := unstructured.NestedString(metadata, "generateName"); err != nil {
+		unstructured.RemoveNestedField(metadata, "generateName")
+	}
+	if _, _, err := unstructured.NestedString(metadata, "selfLink"); err != nil {
+		unstructured.RemoveNestedField(metadata, "selfLink")
+	}
+	if _, _, err := unstructured.NestedString(metadata, "uid"); err != nil {
+		unstructured.RemoveNestedField(metadata, "uid")
+	}
+	if _, _, err := unstructured.NestedInt64(metadata, "generation"); err != nil {
+		unstructured.RemoveNestedField(metadata, "generation")
+	}
+	if _, _, err := unstructured.NestedInt64(metadata, "deletionGracePeriodSeconds"); err != nil {
+		unstructured.RemoveNestedField(metadata, "deletionGracePeriodSeconds")
+	}
+	if labels, ok := metadata["labels"].(map[string]interface{}); !ok {
+		unstructured.RemoveNestedField(metadata, "labels")
+	} else if labels != nil {
+		for k, v := range labels {
+			if _, ok := v.(string); !ok {
+				delete(labels, k)
+			}
+		}
+	}
+	if annotations, ok := metadata["annotations"].(map[string]interface{}); !ok {
+		unstructured.RemoveNestedField(metadata, "annotations")
+	} else if annotations != nil {
+		for k, v := range annotations {
+			if _, ok := v.(string); !ok {
+				delete(annotations, k)
+			}
+		}
+	}
+
+	if finalizers, ok := metadata["finalizers"].([]interface{}); !ok {
+		unstructured.RemoveNestedField(metadata, "finalizers")
+	} else if finalizers != nil {
+		var newFinalizers []interface{}
+		for _, f := range finalizers {
+			if f, ok := f.(string); ok {
+				newFinalizers = append(newFinalizers, f)
+			}
+		}
+		unstructured.SetNestedField(metadata, newFinalizers, "finalizers")
+	}
+	if _, _, err := unstructured.NestedString(metadata, "clusterName"); err != nil {
+		unstructured.RemoveNestedField(metadata, "clusterName")
+	}
+}
+
+func getObjectMeta(u *unstructured.Unstructured) (*metav1.ObjectMeta, error) {
+	metadata := u.UnstructuredContent()["metadata"]
+	if metadata == nil {
+		return nil, nil
+	}
+
+	metadataMap, ok := metadata.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("invalid metadata: expected object, got %T", metadata)
+	}
+
+	objectMeta := &metav1.ObjectMeta{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(metadataMap, objectMeta); err != nil {
```

Comment:
I'm not sure this does all the coercions normal json decoding does (e.g. `null` coerces to string values as "")

https://play.golang.org/p/1HuZBBbmiqQ

see the original issue where we realized this problem existed - https://github.com/kubernetes/kubernetes/issues/59451

if the expectation is that CRDs behave like native resources, we should probably lean on the normal json ObjectMeta decoder as much as possible

Reply:
fixed, and added tests to compare the behaviour of CRDs with a native type and a universal codec.